### PR TITLE
Preview of layers in dxf import dialog

### PR DESCRIPTION
**Pull request by domi4484**
Add a preview when clicking on layers in the dxf/dwg import dialog and some other small UI improvements:
- Automatically fill the main group name line edit using the document file name
- Use `WaitCursor` instead of `BusyCursor` during long lasting operations 

Peek 2023-03-31 11-05

### CODE REVIEW

File: src/app/dwg/qgsdwgimportdialog.cpp [hunk 1]
```diff
@@ -535,3 +539,55 @@ void QgsDwgImportDialog::showHelp()
...
+  QList<QgsMapLayer *> layers = layerGroup.layerOrderRespectingGroupLayers();
+
+  if ( layers.isEmpty() )
+  {
+    mLabelLayerPreview->setText( tr( "No features in this layer" ) );
+    return;
+  }
+
+  // Create image and painterr
+  QImage image( mLabelLayerPreview->size(), QImage::Format_ARGB32_Premultiplied );
```
Comment (nyalldawson) [inline src/app/dwg/qgsdwgimportdialog.cpp:570]: Why not just use a QgsMapCanvas here? That's what the preview feature in browser uses, along with a present pan tool, and it works very nicely.

Comment (domi4484) [reply src/app/dwg/qgsdwgimportdialog.cpp:570]: @nyalldawson thanks, @nirvn already told me the same 🙂. Its for sure much better I'll change it!

File: src/app/dwg/qgsdwgimportdialog.h [hunk 2]
```diff
@@ -41,13 +42,23 @@ class QgsDwgImportDialog : public QDialog, private Ui::QgsDwgImportBase
...
+      Visibility = 1
+    };
+
     QgsVectorLayer *layer( QgsLayerTreeGroup *layerGroup, const QString &layer, const QString &table );
     void createGroup( QgsLayerTreeGroup *group, const QString &name, const QStringList &layers, bool visible );
     void updateUI();
     void expandInserts();
     void updateCheckState( Qt::CheckState state );
+
+    QgsMapToolPan *mPanTool;
```
Comment (nyalldawson) [inline src/app/dwg/qgsdwgimportdialog.h:61]:
```suggestion
    QgsMapToolPan *mPanTool = nullptr;
```

File: src/app/dwg/qgsdwgimportdialog.h [hunk 3]
```diff
@@ -41,13 +42,23 @@ class QgsDwgImportDialog : public QDialog, private Ui::QgsDwgImportBase
     void mDatabaseFileWidget_textChanged( const QString &filename );
     void leLayerGroup_textChanged( const QString &text );
     void showHelp();
+    void layers_clicked( QTableWidgetItem *item );
```
Comment (nyalldawson) [inline src/app/dwg/qgsdwgimportdialog.h:45]:
```suggestion
    void layersClicked( QTableWidgetItem *item );
```

File: src/app/dwg/qgsdwgimportdialog.cpp [hunk 4]
```diff
@@ -535,3 +544,25 @@ void QgsDwgImportDialog::showHelp()
...
+
+  if ( ! item )
+    return;
+
+  const QgsTemporaryCursorOverride waitCursor( Qt::WaitCursor );
+
+  QString layerName = item->text();
+
+  QgsLayerTreeGroup layerGroup;
+  createGroup( &layerGroup, layerName, QStringList( layerName ), true );
```
Comment (nyalldawson) [inline src/app/dwg/qgsdwgimportdialog.cpp:564]: If you follow the chain of function calls this is adding layers to the current project -- not what we want at all for a preview! You'll need to refactor out the layer creation and store the preview layer in a unique_ptr in the dialog only.

**Review by nyalldawson — DISMISSED** [review]

File: src/app/dwg/qgsdwgimportdialog.cpp (hunk 4 above)
Comment (domi4484) [reply src/app/dwg/qgsdwgimportdialog.cpp]: Thank you, I did not see it. I added a new boolean parameter to createGroup and layer functions `addToProject`.

Comment (nyalldawson) [reply src/app/dwg/qgsdwgimportdialog.cpp]: If add to project is false then the layer is leaked. 

I think this needs more extensive refactoring -- the actual layer creation and styling code should be made independent of the layer group handling. Then the preview functionality should just call the layer creator function, store the result in a member unique ptr and set the canvas layer accordingly. There's no need for the preview functionality to handle layer tree related logic at all.

Comment (domi4484) [reply src/app/dwg/qgsdwgimportdialog.cpp]: Make sense thanks. But the layers to preview are multiple, one dxf layer correspond to multiple qgis layers, so I put them on a list and delete it in the destructor

File: src/app/dwg/qgsdwgimportdialog.cpp [hunk 5]
```diff
@@ -443,24 +452,31 @@ void QgsDwgImportDialog::createGroup( QgsLayerTreeGroup *group, const QString &n
...
+  QgsDebugMsg( QStringLiteral( " %1" ).arg( name ) ) ;
+  Q_ASSERT( layerGroup );
+
+  const QList<QgsVectorLayer *> layersList = createLayers( layers );
+  for ( QgsVectorLayer *layer : layersList )
   {
-    l = layer( layerGroup, layerFilter, QStringLiteral( "inserts" ) );
-    if ( l && l->renderer() )
-    {
-      QgsSingleSymbolRenderer *ssr = dynamic_cast<QgsSingleSymbolRenderer *>( l->renderer() );
```
Comment (nyalldawson) [inline src/app/dwg/qgsdwgimportdialog.cpp:460]: Can you explain why this code is no longer required?

File: src/app/dwg/qgsdwgimportdialog.cpp (hunk 4 above)
Comment (nyalldawson) [reply src/app/dwg/qgsdwgimportdialog.cpp]: Looking good, thanks!

File: src/app/dwg/qgsdwgimportdialog.cpp (hunk 5 above)
Comment (domi4484) [reply src/app/dwg/qgsdwgimportdialog.cpp:460]: It is a mistake, I re-added it.

Comment (domi4484) [reply src/app/dwg/qgsdwgimportdialog.cpp:460]: I added a test that checks that those awaited layers are effectively in the project after using the import dialog.
@nyalldawson could you please take another look? I would like to have this into 3.32

**Comment by nyalldawson** [comment]
Looking good to me now! My only questions is whether the removal of the `cbExpandInserts->isChecked()` test will have unwanted side effects?

**Comment by domi4484** [comment]
Oh this wasn't intended 🫣

But in fact the current implementation is not completely logic to me, If the checkbox is checked or not, blocks are expanded anyway. At least is what I am seeing when importing tests/testdata/dwg/entities.dwg. In the picture left I would have expected to see only the points and not the squares. So maybe is just the name of the checkbox wrong and should be named something like "Add inserts layer" instead?

image

But when checked, there is really a function called `expandInserts` that is called [here](https://github.com/qgis/QGIS/blob/master/src/app/dwg/qgsdwgimporter.cpp#L875)...
I need to investigate this a little more, if you have some input or context would be appreciated!

**Comment by nyalldawson** [comment]
@domi4484 
Sorry, I have no idea what this is supposed to do... I'm not a cad guy myself! Maybe @lbartoletti @jef-n @3nids @mhugent have insights?

**Comment by andreasneumann** [comment]
> But in fact the current implementation is not completely logic to me, If the checkbox is checked or not, blocks are expanded anyway. At least is what I am seeing when importing tests/testdata/dwg/entities.dwg. In the picture left I would have expected to see only the points and not the squares. So maybe is just the name of the checkbox wrong and should be named something like "Add inserts layer" instead?

It looks like the difference is whether the original block anchor points are kept or not. The block geometry is expanded in both cases? Maybe something like "Keep block anchor points"? For sure @jef-n or @mhugent could help here.

**Comment by domi4484** [comment]
I brought back the original behavior of `cbExpandInserts` in relations to the `inserts` layer. 
If someone has some background information about this checkbox I am still interested, and if it is meaningful to change something I will be happy to do it in a separate P.R.

**Comment by andreasneumann** [comment]
> I brought back the original behavior of `cbExpandInserts` in relations to the `inserts` layer. If someone has some background information about this checkbox I am still interested, and if it is meaningful to change something I will be happy to do it in a separate P.R.

I think the original behavior was that block references were only displayed as simple points - and not

**Comment by domi4484** [comment]
Thank you @andreasneumann, I like your suggestion and i replaced the check box with a combobox with three modes:
- Expand block geometries (default)
- Expand block geometries and add insert points
- Add only insert points

image

File: src/app/dwg/qgsdwgimportdialog.cpp [hunk 6]
```diff
@@ -57,6 +57,17 @@ QgsDwgImportDialog::QgsDwgImportDialog( QWidget *parent, Qt::WindowFlags f )
   mDatabaseFileWidget->setStorageMode( QgsFileWidget::SaveFile );
   mDatabaseFileWidget->setConfirmOverwrite( false );
 
+  mBlockModeComboBox->addItem( tr( "Expand block geometries" ), static_cast<int>( BlockImportFlag::BlockImportExpandGeometry ) );
+  mBlockModeComboBox->addItem( tr( "Expand block geometries and add insert points" ), static_cast<int>( BlockImportFlag::BlockImportExpandGeometry ) | static_cast<int>( BlockImportFlag::BlockImportAddInsertPoints ) );
+  mBlockModeComboBox->addItem( tr( "Add only insert points" ), static_cast<int>( BlockImportFlag::BlockImportAddInsertPoints ) );
```
Comment (nyalldawson) [inline src/app/dwg/qgsdwgimportdialog.cpp:62]:
```suggestion
  mBlockModeComboBox->addItem( tr( "Expand Block Geometries" ), static_cast<int>( BlockImportFlag::BlockImportExpandGeometry ) );
  mBlockModeComboBox->addItem( tr( "Expand Block Geometries and Add Insert Points" ), static_cast<int>( BlockImportFlag::BlockImportExpandGeometry ) | static_cast<int>( BlockImportFlag::BlockImportAddInsertPoints ) );
  mBlockModeComboBox->addItem( tr( "Add Only Insert Points" ), static_cast<int>( BlockImportFlag::BlockImportAddInsertPoints ) );
```

**Comment by nyalldawson** [comment]
LGTM!

**Review by nirvn — APPROVED** [review]
Looks good to me too, let's merge when this is green. Thanks for the work on this @domi4484 .

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**